File: frontend/src/api/npc.ts
/**
 * NPC API Client
 * --------------
 * Functions for interacting with AI-powered NPCs via the backend.
 */

import { AiNpcProfile, ChatMessage } from '../types/aiNpc';
import { NpcCharacter } from '../types/npcCharacter';
import { buildNpcPrompt } from '../game/data/NpcPromptBuilder';

const API_BASE = '/api/npc';

interface ChatResponse {
  npc_response: string;
}

/**
 * Send a message to an AI NPC and get their response (simple version).
 */
export async function chatWithNpc(
  npc: AiNpcProfile,
  conversationHistory: ChatMessage[],
  playerMessage: string
): Promise<string> {
  try {
    const response = await fetch(`${API_BASE}/chat`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        npc: {
          name: npc.name,
          backstory: npc.backstory,
          personality: npc.personality,
          guidelines: npc.guidelines,
        },
        conversation_history: conversationHistory,
        player_message: playerMessage,
      }),
    });

    if (!response.ok) {
      console.error('NPC chat error:', response.status);
      return `*${npc.name} doesn't respond*`;
    }

    const data: ChatResponse = await response.json();
    return data.npc_response;
  } catch (error) {
    console.error('NPC chat error:', error);
    return `*${npc.name} seems lost in thought*`;
  }
}

/**
 * Send a message to a full NPC character and get their response.
 * Uses the complete character sheet for rich interactions.
 */
export async function chatWithNpcCharacter(
  npc: NpcCharacter,
  conversationHistory: ChatMessage[],
  playerMessage: string
): Promise<string> {
  try {
    // Build the full system prompt from the character sheet
    const systemPrompt = buildNpcPrompt(npc);

    const response = await fetch(`${API_BASE}/chat`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        npc: {
          name: npc.background.name,
        },
        conversation_history: conversationHistory,
        player_message: playerMessage,
        system_prompt: systemPrompt,
      }),
    });

    if (!response.ok) {
      console.error('NPC chat error:', response.status);
      return `*${npc.background.name} doesn't respond*`;
    }

    const data: ChatResponse = await response.json();
    return data.npc_response;
  } catch (error) {
    console.error('NPC chat error:', error);
    return `*${npc.background.name} seems lost in thought*`;
  }
}

File: backend/app/api/routes/npc.py
"""
NPC Chat Routes
---------------
Handles AI-powered NPC conversations using Gemini.
"""

from fastapi import APIRouter
from pydantic import BaseModel

from app.services.gemini_service import generate_npc_response


router = APIRouter(prefix="/npc", tags=["npc"])


class NpcProfile(BaseModel):
    """NPC character definition."""
    name: str
    backstory: str = ""
    personality: str = ""
    guidelines: str = ""


class ChatMessage(BaseModel):
    """A single message in the conversation."""
    role: str  # "player" or "npc"
    content: str


class ChatRequest(BaseModel):
    """Request to chat with an NPC."""
    npc: NpcProfile
    conversation_history: list[ChatMessage] = []
    player_message: str
    system_prompt: str | None = None  # Pre-built prompt (overrides npc fields)


class ChatResponse(BaseModel):
    """NPC's response."""
    npc_response: str


@router.post("/chat", response_model=ChatResponse)
async def chat_with_npc(request: ChatRequest) -> ChatResponse:
    """
    Send a message to an AI NPC and get their response.

    The NPC will respond based on their backstory, personality,
    and conversation guidelines while staying in character.
    """
    response = await generate_npc_response(
        npc_name=request.npc.name,
        backstory=request.npc.backstory,
        personality=request.npc.personality,
        guidelines=request.npc.guidelines,
        conversation_history=[msg.model_dump() for msg in request.conversation_history],
        player_message=request.player_message,
        system_prompt=request.system_prompt,
    )

    return ChatResponse(npc_response=response)
